### PR TITLE
Introduction operationlock around Update

### PR DESCRIPTION
**Pull request by stgraber**
Closes https://github.com/lxc/lxd/issues/10081

### CODE REVIEW

File: lxd/instance/drivers/driver_lxc.go [hunk 1]
```diff
@@ -3958,6 +3958,13 @@ func (d *lxc) CGroupSet(key string, value string) error {
 
 // Update applies updated config.
 func (d *lxc) Update(args db.InstanceArgs, userRequested bool) error {
+	// Setup a new operation
+	op, err := operationlock.CreateWaitGet(d.Project(), d.Name(), operationlock.ActionUpdate, []operationlock.Action{operationlock.ActionRestore}, false, false)
+	if err != nil {
+		return fmt.Errorf("Create container update operation: %w", err)
```
Comment (simondeziel) [inline lxd/instance/drivers/driver_lxc.go:3964]: `s/container/instance/`?

File: lxd/instance/drivers/driver_qemu.go [hunk 2]
```diff
@@ -4140,6 +4140,14 @@ func (d *qemu) Rename(newName string, applyTemplateTrigger bool) error {
 
 // Update the instance config.
 func (d *qemu) Update(args db.InstanceArgs, userRequested bool) error {
+	// Setup a new operation.
+	op, err := operationlock.CreateWaitGet(d.Project(), d.Name(), operationlock.ActionUpdate, []operationlock.Action{operationlock.ActionRestore}, false, false)
+	if err != nil {
+		return fmt.Errorf("Create container update operation: %w", err)
```
[the remark this answers is not in the document]
Comment (tomponline) [reply lxd/instance/drivers/driver_qemu.go:4146]: My vote would be instance for both

File: lxd/instance/drivers/driver_lxc.go [hunk 3]
```diff
@@ -3958,6 +3958,13 @@ func (d *lxc) CGroupSet(key string, value string) error {
 
 // Update applies updated config.
 func (d *lxc) Update(args db.InstanceArgs, userRequested bool) error {
+	// Setup a new operation
+	op, err := operationlock.CreateWaitGet(d.Project(), d.Name(), operationlock.ActionUpdate, []operationlock.Action{operationlock.ActionRestore}, false, false)
+	if err != nil {
+		return fmt.Errorf("Create instance update operation: %w", err)
```
Comment (tomponline) [inline lxd/instance/drivers/driver_lxc.go:3964]: Can you use the form "Failed to..." So it reads better in a client facing error?

File: lxd/instance/drivers/driver_lxc.go [hunk 4]
```diff
@@ -3958,6 +3958,13 @@ func (d *lxc) CGroupSet(key string, value string) error {
 
 // Update applies updated config.
 func (d *lxc) Update(args db.InstanceArgs, userRequested bool) error {
+	// Setup a new operation
+	op, err := operationlock.CreateWaitGet(d.Project(), d.Name(), operationlock.ActionUpdate, []operationlock.Action{operationlock.ActionRestore}, false, false)
```
Comment (tomponline) [inline lxd/instance/drivers/driver_lxc.go:3962]: Why action restore here BTW?

**Review by tomponline — APPROVED** [review]

File: lxd/instance/drivers/driver_lxc.go (hunk 4 above)
Comment (stgraber) [reply lxd/instance/drivers/driver_lxc.go:3962]: Because the Restore function calls Update() and we don't want that to deadlock.

**Review by tomponline — APPROVED** [review]
Maybe link to the issue this fixes in pr description?

**Comment by stgraber** [comment]
@tomponline good to go

**Pull request merged**